src/utils/async.ts

```diff
@@ -30,6 +30,61 @@ export class ManualPromise {
     }
 }
 
+/**
+ * WaitGroup tracks the progress of multiple asynchronous tasks, allowing a caller to wait for all tasks to complete.
+ *
+ * Before each task is spawned, a consumer will call `add` for each spawned task, to indicate that there is a new task
+ * to wait for. Each task independently marks itself as done with the `done` method. Once the number of `done` calls
+ * equals the number of `add` calls, the promise resolves, and a consumer will know the tasks are complete.
+ *
+ * This is useful in contrast to `Promise.all`/`Promise.allSettled`, where it may not be easily determinable up front
+ * how many tasks you are waiting on (e.g. if you wish all event handler functions for event A to fully complete
+ * before an event handler for event B runs).
```

Comment:
It's a bit easier to understand in context. When the `HighlightManager` gets a redraw event, we add to the WaitGroup before deferring on an async task. Then, if while waiting for that async task, we get `redraw-flush`, we want to make sure that `redraw` event has completed (i.e. said async task has finished) before continuing. 

Your example doesn't fit this bill because if we add to `listA` once `Promise.all` has been called, adding items to listA will not prolong the resolution of `Promise.all` (at least, as far as I know). This structure allows for that